File: Array/1.Easy/11.Max_consecutive_1's.cpp
/*
QUESTION:-
Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.

Example 1:

Input: nums = [3,0,1]
Output: 2
Explanation: n = 3 since there are 3 numbers, so all numbers are in the range [0,3]. 2 is the missing number in the range since it does not appear in nums.
Example 2:

Input: nums = [0,1]
Output: 2
Explanation: n = 2 since there are 2 numbers, so all numbers are in the range [0,2]. 2 is the missing number in the range since it does not appear in nums.
*/


#include <bits/stdc++.h>
using namespace std;
/*
APPROACH:-
-> Traverse the entire array and within it run a loop while element's are equal to 1 and store the count
-> Update the ans as max(ans,cnt)
*/

// CODE:-
int findMaxConsecutiveOnes(vector<int> &nums)
{
    int ans = 0;
    for (int i = 0; i < nums.size(); i++)
    {
        int cnt = 0; // to store the count of consecutive 1's
        while (i < nums.size() && nums[i] == 1)
        {
            cnt++;
            i++;
        }
        ans = max(ans, cnt);
    }
    return ans;
}


int main() {
    // Test case 1
    cout << "Test 1:" << endl;
    vector<int> nums1 = {3,0,1};
    int result1 = findMaxConsecutiveOnes(nums1);
    cout << "Output: " << result1 << endl;
    cout << "Expected: 2" << endl;
    cout << endl;

    // Test case 2
    cout << "Test 2:" << endl;
    vector<int> nums2 = {0,1};
    int result2 = findMaxConsecutiveOnes(nums2);
    cout << "Output: " << result2 << endl;
    cout << "Expected: 2" << endl;
    cout << endl;

    return 0;
}

// TIME COMPLEXITY = O(N)

File: 01.Arrays/1.Easy/11.Max_consecutive_1's.py
# QUESTION:-
# Find the maximum count of consecutive 1's in a given integer array nums containing only 0's and 1's.
# Example 1: Input: [1,1,0,1,1,1], Output: 3
# Example 2: Input: [1,0,1,1,0,1], Output: 2

# APPROACH:-
# Traverse the entire array
# Within it run a loop while elements are equal to 1 and store the count
# Update the ans as max(ans, cnt)

# CODE:-
def findMaxConsecutiveOnes(nums):
    """Find the maximum count of consecutive 1's"""
    ans = 0
    i = 0
    while i < len(nums):
        cnt = 0  # to store the count of consecutive 1's
        while i < len(nums) and nums[i] == 1:
            cnt += 1
            i += 1
        ans = max(ans, cnt)
        if i < len(nums):
            i += 1  # move past the 0
    return ans


if __name__ == '__main__':
    # Test case 1
    print("Test 1:")
    nums1 = [1, 1, 0, 1, 1, 1]
    result1 = findMaxConsecutiveOnes(nums1)
    print(f"Output: {result1}")
    print("Expected: 3")
    print()
    
    # Test case 2
    print("Test 2:")
    nums2 = [1, 0, 1, 1, 0, 1]
    result2 = findMaxConsecutiveOnes(nums2)
    print(f"Output: {result2}")
    print("Expected: 2")

# TIME COMPLEXITY = O(N)
# SPACE COMPLEXITY = O(1)
